hibernate.go

```diff
@@ -0,0 +1,330 @@
+	indexDefs, _, err := mgr.GetIndexDefs(true)
+	if err != nil {
+		return result, errors.Wrapf(err, "error getting index defs")
+	}
+	if indexDefs != nil {
+		for _, index := range indexDefs.IndexDefs {
+			bucketName := index.SourceName
+			indexName := index.Name
+			path := bucketName + ":" + indexName + ":" + LastAccessedTime
+			resByte, _, _, err := jp.Get(data, path)
+			if err != nil {
+				return result, errors.Wrapf(err, "error extracting data from response")
+			}
+			stringRes := string(resByte)
+			if stringRes != "" {
+				resTime, err := time.Parse(time.RFC3339, stringRes)
+				if err != nil {
+					log.Printf("Hibernate: error parsing result into time format: %e", err)
+				} else {
+					result[indexName] = individualIndexActivityStats{
+						LastAccessedTime: resTime,
+					}
+				}
+			}
+		}
+	}
+	return result, nil
+}
+
+// Returns a FQDN URL to the /nsstats endpoint.
+func getNsStatsURL(mgr *cbgt.Manager, bindHTTP string) (string, error) {
+	cfg := mgr.Cfg()
+
+	nodeDefs, _, err := cbgt.CfgGetNodeDefs(cfg, cbgt.NODE_DEFS_KNOWN)
+	if err != nil {
+		return "", errors.Wrapf(err, "Error getting nodedefs")
+	}
+	if nodeDefs == nil {
+		return "", fmt.Errorf("empty nodedefs")
+	}
+
+	prefix := mgr.Options()["urlPrefix"]
+	// polling one node endpoint is enough - avoids duplicate statistics.
+	for _, node := range nodeDefs.NodeDefs {
+		fqdnURL := "http://" + bindHTTP + prefix + "/api/nsstats"
+		// no auth required for /nsstats
+		secSettings := cbgt.GetSecuritySetting()
+		if secSettings.EncryptionEnabled {
+			if httpsURL, err := node.HttpsURL(); err == nil {
+				fqdnURL = httpsURL
+			}
+		}
+		// request error: parse 192.168.1.6:9200/api/nsstats: first path segment in URL cannot contain colon
+		// Ref: https://stackoverflow.com/questions/54392948/first-path-segment-in-url-cannot-contain-colon
+		fqdnURL = strings.Trim(fqdnURL, "\n")
+		// possibly check node health here?
+		// return only if node is healthy?
+		return fqdnURL, nil
+	}
+	return "", nil
+}
+
+// IndexHibernateProbe monitors an index's activity.
+func IndexHibernateProbe(mgr *cbgt.Manager, bindHTTP string) error {
+	url, err := getNsStatsURL(mgr, bindHTTP)
+	if err != nil {
+		return errors.Wrapf(err, "Error getting /nsstats endpoint URL")
+	}
+
+	indexActivityStatsSample := make(chan indexActivityStats)
+	options := &indexActivityStatsOptions{
+		indexActivityStatsSampleInterval: 10 * time.Second,
+	}
+
+	resCh, err := startIndexActivityMonitor(url, indexActivityStatsSample, *options)
+	if err != nil {
+		return errors.Wrapf(err, "Error starting NSMonitor")
+	}
+	go func() {
+		for r := range resCh.SampleCh {
+			result, err := unmarshalIndexActivityStats(r.Data, mgr)
+			if err != nil {
+				log.Printf("Hibernate: Error unmarshalling stats: %e", err)
+				continue
+				// should continue even if unmarshalling does not
+				// work for one endpoint.
+			}
+			//log.Printf("Hibernate: Result: %+v", result)
+			go updateHibernationStatus(mgr, result)
```

Comment:
a routine on every sample collection looks an overhead here? samples could be queued and processed serially?

Reply:
We need to guard against every other node in the cluster doing the same hibernation policy application at the same time/doing redundant work?

Reply:
Regarding redundant work by nodes, I think this check returns the nodeURL for the first node possible:
https://github.com/metonymic-smokey/cbft/blob/6fe3f016aab83a85b04f7000bc328a5b4db4d783/hibernate.go#L198-L212

Reply:
But this code runs on every node in the cluster? So, every node takes the first node and ask it and do the get/update/set operations on the index definitions..?

Reply:
Thanks. In a recent commit, I have added a check which only returns the URL of the manager node.